storage/shard_insert_queue.go

```diff
@@ -48,24 +51,60 @@ type dbShardInsertQueue struct {
 
 	currBatch    *dbShardInsertBatch
 	notifyInsert chan struct{}
+
+	metrics dbShardInsertQueueMetrics
+}
+
+type dbShardInsertQueueMetrics struct {
+	insertsNoPendingWrite tally.Counter
+	insertsPendingWrite   tally.Counter
+}
+
+func newDBShardInsertQueueMetrics(
+	scope tally.Scope,
+) dbShardInsertQueueMetrics {
+	insertName := "inserts"
+	insertPendingWriteTagName := "pending-write"
+	return dbShardInsertQueueMetrics{
+		insertsNoPendingWrite: scope.Tagged(map[string]string{
+			insertPendingWriteTagName: "no",
+		}).Counter(insertName),
+		insertsPendingWrite: scope.Tagged(map[string]string{
+			insertPendingWriteTagName: "yes",
+		}).Counter(insertName),
+	}
 }
 
 type dbShardInsertBatch struct {
 	wg      *sync.WaitGroup
-	entries []*dbShardEntry
+	inserts []dbShardInsert
+}
+
+type dbShardInsert struct {
+	entry           *dbShardEntry
+	hasPendingWrite bool
+	pendingWrite    dbShardPendingWrite
+}
+
+type dbShardPendingWrite struct {
+	timestamp  time.Time
+	value      float64
+	unit       xtime.Unit
+	annotation []byte
 }
 
 func (b *dbShardInsertBatch) reset() {
 	b.wg = &sync.WaitGroup{}
 	// We always expect to be waiting for an insert
 	b.wg.Add(1)
-	for i := range b.entries {
-		b.entries[i] = nil
+	insertZeroed := dbShardInsert{pendingWrite: dbShardPendingWrite{}}
```

Comment:
How often does `reset` get called? Should we perhaps make the empty `dbShardInsert` object a file-scope `var`?